src/logger.hpp

```diff
@@ -0,0 +1,164 @@
+    {
+        auto android_log_level = map_android_log_level[entry.first];
+        __android_log_print(android_log_level, "realm", "%s", entry.second.c_str());
+    }
+};
+#endif
+
+#ifdef __APPLE__
+class IOSLogger {
+    void print() {}
+};
+#endif
+
+class SyncLogger : public realm::util::RootLogger {
+public:
+    void delegate(Delegated&& delegate)
+    {
+        m_scheduler->set_notify_callback([this, delegate = std::move(delegate)] {
+            while (!m_log_queue.empty()) {
+                /*
+                 *  Extracting and delegating log entries.
+                 */
+                Entry entry;
+                {
+                    std::lock_guard<std::mutex> lock(m_mutex);
+                    if (m_log_queue.empty())
+                        return;
+                    entry = m_log_queue.front();
+                    m_log_queue.pop();
+                }
+
+                delegate(static_cast<int>(entry.first), entry.second);
+            }
+        });
+    }
+
+protected:
+    void do_log(LoggerLevel level, std::string message)
+    {
+        std::lock_guard<std::mutex> lock(m_mutex);
+
+        // TODO we are coupling core with JS here, change to string use hashmap
+        // map_level.
+        auto entry = std::make_pair(level, message);
+
+        m_log_queue.push(entry);
+        m_scheduler->notify();
+    }
+
+private:
+    std::queue<Entry> m_log_queue;
+    Delegated loggerDelegate;
+    std::shared_ptr<realm::util::Scheduler> m_scheduler = realm::util::Scheduler::make_default();
+    std::mutex m_mutex;
+};
+
+class LoggerFactory : public realm::SyncLoggerFactory {
+public:
+    LoggerFactory(Delegated&& logs_fn)
+        : logs_fn{logs_fn}
+    {
+    }
+
+    std::unique_ptr<realm::util::Logger> make_logger(realm::util::Logger::Level level)
+    {
+        auto logger = std::make_unique<SyncLogger>();
+
+        logger->set_level_threshold(level);
+        logger->delegate(std::move(logs_fn));
+
+        return logger;
+    }
+
+private:
+    Delegated logs_fn;
+};
+
+class Logger {
+private:
+    /*
+       Log levels available.
+       More info in (realm-core) realm/util/logger.hpp
+
+       [ all, trace, debug, detail, info, warn, error, fatal, off ]
+    */
+    std::map<LoggerLevel, std::string> map_level{
+        {LoggerLevel::all, "all"},     {LoggerLevel::info, "info"},     {LoggerLevel::trace, "trace"},
+        {LoggerLevel::debug, "debug"}, {LoggerLevel::detail, "detail"}, {LoggerLevel::warn, "warn"},
+        {LoggerLevel::error, "error"}, {LoggerLevel::fatal, "fatal"},   {LoggerLevel::off, "off"},
+    };
+
+public:
+    LoggerLevel get_level(std::string level)
+    {
+        for (auto const& [key, value] : map_level) {
+            if (value == level)
+                return key;
+        }
+
+        throw std::runtime_error("Bad log level");
+    }
+
+    LoggerFactory* build_sync_logger(Delegated&& log_fn)
+    {
+        return new LoggerFactory{std::move(log_fn)};
+    }
+};
```

Comment:
Is there any particular reason why the two functions above aren't `static`?  That would eliminate the need to instantiate a `Logger` class to retrieve a `LoggerFactory`.